flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierTracker.java

```diff
@@ -82,7 +82,9 @@ public void processBarrier(CheckpointBarrier receivedBarrier, InputChannelInfo c
 
 		// fast path for single channel trackers
 		if (totalNumberOfInputChannels == 1) {
-			notifyCheckpoint(receivedBarrier, 0);
+			markCheckpointStart(receivedBarrier.getTimestamp());
+			markAlignmentEnd(0);
```

Comment:
I'm not sure if I'm following your suggested change? If you ignore how are those methods implemented under the hood, those calls here make sense.

We are firt marking start of the checkpoint, and we are marking that it finished immediately, with the duration of `0`. (note that the first call is also setting the `checkpointStartDelay`).

This is the only place, where those two calls are called one beside another, if I would extracted this to a separate method, that would be a slight overkill.

Now about how are they implemented under the hood, it's done in order to avoid a bit of code duplication.

Reply:
I'd just decouple the two metrics of checkpoint start duration and alignment times. For tracker, there is no semantics of alignment. We have an extra call to `System.nanoTime()` that is completely unnecessary (here I don't worry about performance). It feels like a mismatch of abstraction. Similarly, when you optimize 1 channel cases, there is just no alignment happening, which you simulate by ending it early and explicitly overwrite the measurement.